benchmark/reconstruction/evaluate.py

```diff
@@ -0,0 +1,974 @@
+# POSSIBILITY OF SUCH DAMAGE.
+
+import argparse
+import collections
+import copy
+import datetime
+import functools
+import multiprocessing
+import shutil
+import subprocess
+from pathlib import Path
+
+import numpy as np
+from PIL import Image
+
+import pycolmap
+
+
+SceneInfo = collections.namedtuple(
+    "SceneInfo",
+    [
+        "category",
+        "scene",
+        "workspace_path",
+        "image_path",
+        "sparse_gt_path",
+        "camera_priors_from_sparse_gt",
+        "position_accuracy_gt",
+        "colmap_extra_args",
+    ],
+)
+
+SceneResult = collections.namedtuple(
+    "SceneResult",
+    [
+        "scene_info",
+        "errors",
+        "num_images",
+        "num_reg_images",
+        "num_components",
+        "largest_component",
+    ],
+)
+
+SceneMetrics = collections.namedtuple(
+    "ScenSceneMetricsResult",
+    [
+        "aucs",
+        "num_images",
+        "num_reg_images",
+        "num_components",
+        "largest_component",
+    ],
+)
+
+
+def update_camera_priors_from_sparse_gt(
+    database_path: Path, camera_priors_sparse_gt: pycolmap.Reconstruction
+) -> None:
+    print("Setting prior cameras from GT")
+
+    database = pycolmap.Database()
+    database.open(database_path)
+
+    camera_id_gt_to_camera_id = {}
+    for camera_id_gt, camera_gt in camera_priors_sparse_gt.cameras.items():
+        camera_gt.has_prior_focal_length = True
+        camera_id = database.write_camera(camera_gt)
+        camera_id_gt_to_camera_id[camera_id_gt] = camera_id
+
+    images_gt_by_name = {}
+    for image_gt in camera_priors_sparse_gt.images.values():
+        images_gt_by_name[image_gt.name] = image_gt
+
+    for image in database.read_all_images():
+        if image.name not in images_gt_by_name:
+            print(
+                f"Not setting prior camera for image {image.name}, "
+                "because it does not exist in GT"
+            )
+            continue
+        image_gt = images_gt_by_name[image.name]
+        camera_id = camera_id_gt_to_camera_id[image_gt.camera_id]
+        image.camera_id = camera_id
+        database.update_image(image)
+
+    database.close()
+
+
+def colmap_reconstruction(
+    args: argparse.Namespace,
+    workspace_path: Path,
+    image_path: Path,
+    camera_priors_sparse_gt: pycolmap.Reconstruction = None,
+    colmap_extra_args: list = None,
+    num_threads: int = 1,
+) -> None:
+    workspace_path.mkdir(parents=True, exist_ok=True)
+
+    database_path = workspace_path / "database.db"
+    if args.overwrite_database and database_path.exists():
+        database_path.unlink()
+
+    sparse_path = workspace_path / "sparse"
+    if args.overwrite_reconstruction and sparse_path.exists():
+        shutil.rmtree(sparse_path)
+
+    if sparse_path.exists():
+        print("Skipping reconstruction, as it already exists")
+        return
+
+    colmap_args = [
+        args.colmap_path,
+        "automatic_reconstructor",
```

Comment:
We should ultimately bind this call such that we can use this benchmark to check that the python vs command line interfaces produce consistent results.

Reply:
Agreed conceptually. One problem is though that we currently do not produce pycolmap packages with CUDA.

Reply:
Let's track this separately. I'll add a TODO here.